Check GCC version on Ubuntu

sdmonov opened the pull request:
GCC 5 in Ubuntu 17.10 and newer doesn't define the macro _GLIBCXX_USE_C99
and causes std::to_string, std::isnan, std::isinf (and more) functions
not to be defined neither. This fix checks if GCC 5 is used on Ubuntu 17.10
or later and shows an error message describing the problem.

Fixes #5229

sdmonov committed:
Check GCC version on Ubuntu

GCC 5 in Ubuntu 17.10 and newer doesn't define the macro _GLIBCXX_USE_C99
and causes std::to_string, std::isnan, std::isinf (and more) functions
not to be defined neither. This fix checks if GCC 5 is used on Ubuntu 17.10
or later and shows an error message describing the problem.

soumith commented on aten/CMakeLists.txt:
```diff
@@ -10,6 +10,34 @@ project(ATen)
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "6.0.0"
+          AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS "5.0.0")
+            message(FATAL_ERROR
+              "Do not use GCC 5. GCC 5 has a known bug in Ubuntu 17.04"
```
shouldn't this be `17.10` and not `17.04`?

sdmonov replied:
You are right. It was a typo. Fixed it.

sdmonov committed:
Check GCC version on Ubuntu

GCC 5 in Ubuntu 17.10 and newer doesn't define the macro _GLIBCXX_USE_C99
and causes std::to_string, std::isnan, std::isinf (and more) functions
not to be defined neither. This fix checks if GCC 5 is used on Ubuntu 17.10
or later and shows an error message describing the problem.

Fixes #5229

ezyang commented:
It is fairly easy for this cmake test to misfire and prevent you from building ATen when you would have succeeded. All you need is to:

1. Run Ubuntu 17.04
2. Have manually built / procured a copy of GCC 5 which doesn't have the bug

It would have been far better to have added a `CHECK_CXX_SOURCE_COMPILES` test for the missing symbols and fail if this compile test fails.
